### DISCUSSION TITLE
Enum values wrongly converted

### DISCUSSION THREAD

**Discussion by Garfield20001** [RicoSuter/NSwag#4490] [Q&A]
Dear all,
I have the problem that NSwagStudio converts enums in a swagger wrong. Or the other way round. How do I have to define a swagger that a enum is converted the right way?

Here is my swagger:
{
	"openapi": "3.0.1",
	"components": {
		"schemas": {			
			"FluxModels.Base.UnitLevel": {
				"enum": [
					"Package",
					"Bundle",
					"Mastercase",
					"Pallet"
				],
				"type": "string",
				"x-enumNames": [
					"Package",
					"Bundle",
					"Mastercase",
					"Pallet"
				],
				"x-ms-enum": {
					"name": "UnitLevel",
					"modelAsString": false,
					"values": [
						{
							"value": 1,
							"name": "Package"
						},
						{
							"value": 4,
							"name": "Bundle"
						},
						{
							"value": 7,
							"name": "Mastercase"
						},
						{
							"value": 10,
							"name": "Pallet"
						}
					]
				}
			}
		}
	}
}

And that is what I get as C# code:

namespace DataAccessLibrary.API
{
    using System = global::System;

    [System.CodeDom.Compiler.GeneratedCode("NJsonSchema", "13.19.0.0 (NJsonSchema v10.9.0.0 (Newtonsoft.Json v13.0.0.0))")]
    public enum UnitLevel
    {

        [System.Runtime.Serialization.EnumMember(Value = @"Package")]
        Package = 0,

        [System.Runtime.Serialization.EnumMember(Value = @"Bundle")]
        Bundle = 1,

        [System.Runtime.Serialization.EnumMember(Value = @"Mastercase")]
        Mastercase = 2,

        [System.Runtime.Serialization.EnumMember(Value = @"Pallet")]
        Pallet = 3,

    }
}

As you can see in the enum of the C# class e.g. the value of a Pallet is 3 but it should be 10. 
How can I convince NSwagStudio to do it the right way?
Or how must the swagger look like to get the expected result?

Thanks for help!!

**Comment by olegd-superoffice** [comment]
If I understand this right, you need to serialize your enum as integer and provide `x-enumNames` attribute.

**Comment by Garfield20001** [comment]
The enumname attribute is not the problem. In the swagger description the value is set to 1 although it is the first enum value in the list (and for an int enum in C# this would normally assigned to zero)

          "printerSupplier": {
            "allOf": [
              {
                "$ref": "#/components/schemas/FluxModels.ManScan.PrinterSupplier"
              }
            ],
            "description": "Supplying company of the printer",
            "example": "",
            "x-enumNames": [
              "Zebra"
            ],
            "x-ms-enum": {
              "name": "PrinterSupplier",
              "modelAsString": false,
              "values": [
                {
                  "value": 1,
                  "name": "Zebra"
                }
              ]
            }
          },

What I need is that in the resulting C# client that I get an enum the following way:

   /// <summary>
   /// Supplier of label printer
   /// </summary>
   [System.CodeDom.Compiler.GeneratedCode("NJsonSchema", "13.18.0.0 (NJsonSchema v10.8.0.0 (Newtonsoft.Json v13.0.0.0))")]
   public enum PrinterSupplier
   {

       [System.Runtime.Serialization.EnumMember(Value = @"Zebra")]
       Zebra = 1,

   }

and not the wrong way (like it currently is)

    /// <summary>
    /// Supplier of label printer
    /// </summary>
    [System.CodeDom.Compiler.GeneratedCode("NJsonSchema", "13.20.0.0 (NJsonSchema v10.9.0.0 (Newtonsoft.Json v13.0.0.0))")]
    public enum PrinterSupplier
    {

        [System.Runtime.Serialization.EnumMember(Value = @"Zebra")]
        Zebra = 0,

    }

Thanks for help!!

René

Von: olegd-superoffice ***@***.***>
Gesendet: Montag, 21. August 2023 12:26
An: RicoSuter/NSwag ***@***.***>
Cc: René Petton ***@***.***>; Author ***@***.***>
Betreff: Re: [RicoSuter/NSwag] Enum values wrongly converted (Discussion #4490)


If I understand this right, you need to serialize your enum as integer and provide x-enumNames attribute.
